### PR TITLE
Adding support for configuring Metal LB via Custom Resources (necessary for 0.13+)

### LINKED ISSUE #322: Add MetalLB 0.13 support
It looks like the latest MetalLB releases aren’t working with the CCM anymore. I suspect the reason is there’s updated syntax with the newer MetalLB 0.13 releases and the CCM isn’t updated to support it.

With CCM 3.5.0 and MetalLB v0.13.4 I get this error on the MetalLB speaker pods:
```
{"caller":"main.go:282","error":"assigned IP not allowed by config","ips":["86.109.11.206"],"level":"error","msg":"IP allocated by controller not allowed by config","op":"setBalancer","ts":"2022-08-30T17:12:23Z"}
```
When I revert MetalLB to 0.12.1, it all works fine.

To replicate the issue:
1. Deploy Kubernetes cluster.
2. Install CCM 3.5.0.
3. Install MetalLB v0.13.4 as shown [here](https://metallb.universe.tf/installation/#installation-by-manifest). I'm using the native yaml deployment.

### PR DESCRIPTION
**Pull request by cprivitere**
Fixes #322

### CODE REVIEW

**Review by cprivitere — COMMENTED** [review]
Refactor configuration to be pluggable and set the configuration file to be one of those pluggable configurations.

File: metal/loadbalancers/metallb/cr.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+var _ Configurer = (*CRDConfigurer)(nil)
+
+func (*CRDConfigurer) AddPeerByService(add *Peer, svcNamespace, svcName string) bool {
+	return false
+}
+func (*CRDConfigurer) RemovePeersByService(svcNamespace, svcName string) bool { return false }
+func (*CRDConfigurer) RemovePeersBySelector(remove *NodeSelector) bool        { return false }
+func AddAddressPool(add *AddressPool) bool                                    { return false }
```
Comment (displague) [inline metal/loadbalancers/metallb/cr.go:24]: this method should not be here.

**Comment by ocobleseqx** [comment]
In the absence of unit tests, I have created a guide to reproduce and test the metallb 0.13+ support https://gist.github.com/ocobleseqx/3f53ab3e75f03a60cee7f6053ce46280

For testing 
 # TEST 1 - IP ADDRESS POOLS: Create Service and check load balancing
```zsh
ocobleseqx test %kubectl get svc                                 
NAME         TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
kubernetes   ClusterIP   10.96.0.1    <none>        443/TCP   7d22h
ocobleseqx test %kubectl expose deployment nginx-hello-deployment --type=LoadBalancer --name=svc-1

service/svc-1 exposed
ocobleseqx test %kubectl get svc svc-1 -o jsonpath='{.spec.loadBalancerIP}'

147.28.154.166%                                                                                  
ocobleseqx test %kubectl -n metallb-system get ipaddresspools.metallb.io default.svc-1 -o jsonpath='{.spec.addresses[0]}' | cut -d/ -f1

147.28.154.166
ocobleseqx test %kubectl -n metallb-system get bgppeers.metallb.io

NAME                                ADDRESS         ASN     BFD PROFILE   MULTI HOPS
metal-multiarch-k8s-x86-blue-00-0   169.254.255.1   65530                 
metal-multiarch-k8s-x86-blue-00-1   169.254.255.2   65530                 
metal-multiarch-k8s-x86-blue-01-0   169.254.255.1   65530                 
metal-multiarch-k8s-x86-blue-01-1   169.254.255.2   65530                 
ocobleseqx test %kubectl get bgpadvertisements.metallb.io/equinix-metal-bgp-adv -n metallb-system -o json | jq -r ' {name:.metadata.name, pools:.spec.ipAddressPools}'
{
  "name": "equinix-metal-bgp-adv",
  "pools": [
    "default.svc-1"
  ]
}
ocobleseqx test %kubectl get pods -o wide

NAME                                     READY   STATUS    RESTARTS   AGE   IP               NODE                              NOMINATED NODE   READINESS GATES
nginx-hello-deployment-8d7c96f8c-9r4xl   1/1     Running   0          23h   192.168.200.15   metal-multiarch-k8s-x86-blue-01   <none>           <none>
nginx-hello-deployment-8d7c96f8c-b6ztm   1/1     Running   0          23h   192.168.200.14   metal-multiarch-k8s-x86-blue-01   <none>           <none>
nginx-hello-deployment-8d7c96f8c-dl9rl   1/1     Running   0          22h   192.168.200.17   metal-multiarch-k8s-x86-blue-01   <none>           <none>
nginx-hello-deployment-8d7c96f8c-pbmsf   1/1     Running   0          22h   192.168.200.16   metal-multiarch-k8s-x86-blue-01   <none>           <none>
nginx-hello-deployment-8d7c96f8c-w56lv   1/1     Running   0          22h   192.168.200.18   metal-multiarch-k8s-x86-blue-01   <none>           <none>
ocobleseqx test %
ocobleseqx test %for i in {1..5}; do curl -s $(kubectl get svc svc-1 -o jsonpath='{.spec.loadBalancerIP}') | grep -E -o "([0-9]{1,3}[\.]){3}[0-9]{1,3}"; done
curl -s $(kubectl get svc svc-1 -o jsonpath='{.spec.loadBalancerIP}') | grep -E -o "([0-9]{1,3}[\.]){3}[0-9]{1,3}"
192.168.200.16
192.168.200.16
192.168.200.18
192.168.200.18
192.168.200.17
```

 # TEST 2 - BGP PEERS: Remove/Add peers
```zsh
ocobleseqx test %DEVICE_ID=a11f2ef5-8b56-4f12-98a8-12983ad39c9e            
ocobleseqx test %kubectl -n metallb-system get bgppeers.metallb.io
NAME                                ADDRESS         ASN     BFD PROFILE   MULTI HOPS
metal-multiarch-k8s-x86-blue-00-0   169.254.255.1   65530                 
metal-multiarch-k8s-x86-blue-00-1   169.254.255.2   65530                 
metal-multiarch-k8s-x86-blue-01-0   169.254.255.1   65530                 
metal-multiarch-k8s-x86-blue-01-1   169.254.255.2   65530                 
ocobleseqx test %metal device stop -i $DEVICE_ID
Device a11f2ef5-8b56-4f12-98a8-12983ad39c9e successfully stopped.
ocobleseqx test %kubectl get nodes -o wide                        
NAME                                     STATUS     ROLES           AGE     VERSION   INTERNAL-IP   EXTERNAL-IP    OS-IMAGE             KERNEL-VERSION      CONTAINER-RUNTIME
metal-multiarch-k8s-controller-primary   Ready      control-plane   7d22h   v1.24.7   10.65.92.1    147.75.47.51   Ubuntu 18.04.6 LTS   5.4.0-135-generic   containerd://1.5.9
metal-multiarch-k8s-x86-blue-00          NotReady   <none>          7d22h   v1.24.7   10.65.92.5    147.28.148.3   Ubuntu 18.04.6 LTS   5.4.0-136-generic   containerd://1.5.9
metal-multiarch-k8s-x86-blue-01          Ready      <none>          7d22h   v1.24.7   10.65.92.3    139.178.85.1   Ubuntu 18.04.6 LTS   5.4.0-135-generic   containerd://1.5.9
ocobleseqx test %kubectl -n metallb-system get bgppeers.metallb.io

NAME                                ADDRESS         ASN     BFD PROFILE   MULTI HOPS
metal-multiarch-k8s-x86-blue-01-0   169.254.255.1   65530                 
metal-multiarch-k8s-x86-blue-01-1   169.254.255.2   65530                 
ocobleseqx test %metal device start -i $DEVICE_ID
Device a11f2ef5-8b56-4f12-98a8-12983ad39c9e successfully started.
ocobleseqx test %ssh -i metal-key root@147.28.148.3
root@metal-multiarch-k8s-x86-blue-00:~# swapoff -a
root@metal-multiarch-k8s-x86-blue-00:~# systemctl restart kubelet
root@metal-multiarch-k8s-x86-blue-00:~# exit
logout
Connection to 147.28.148.3 closed.
ocobleseqx test %kubectl get nodes -o wide                        
NAME                                     STATUS   ROLES           AGE     VERSION   INTERNAL-IP   EXTERNAL-IP    OS-IMAGE             KERNEL-VERSION      CONTAINER-RUNTIME
metal-multiarch-k8s-controller-primary   Ready    control-plane   7d22h   v1.24.7   10.65.92.1    147.75.47.51   Ubuntu 18.04.6 LTS   5.4.0-135-generic   containerd://1.5.9
metal-multiarch-k8s-x86-blue-00          Ready    <none>          7d22h   v1.24.7   10.65.92.5    147.28.148.3   Ubuntu 18.04.6 LTS   5.4.0-136-generic   containerd://1.5.9
metal-multiarch-k8s-x86-blue-01          Ready    <none>          7d22h   v1.24.7   10.65.92.3    139.178.85.1   Ubuntu 18.04.6 LTS   5.4.0-135-generic   containerd://1.5.9
ocobleseqx test %kubectl -n metallb-system get bgppeers.metallb.io

NAME                                ADDRESS         ASN     BFD PROFILE   MULTI HOPS
metal-multiarch-k8s-x86-blue-00-0   169.254.255.1   65530                 
metal-multiarch-k8s-x86-blue-00-1   169.254.255.2   65530                 
metal-multiarch-k8s-x86-blue-01-0   169.254.255.1   65530                 
metal-multiarch-k8s-x86-blue-01-1   169.254.255.2   65530                 
ocobleseqx test %kubectl delete svc svc-1 
service "svc-1" deleted
ocobleseqx test %kubectl -n metallb-system get ipaddresspools.metallb.io
No resources found in metallb-system namespace.
ocobleseqx test %kubectl -n metallb-system get bgpadvertisements.metallb.io
kubectl -n metallb-system get bgppeers.metallb.io
No resources found in metallb-system namespace.
No resources found in metallb-system namespace.
```

 # TEST 3 - BGP PEERS: Remove/Add peers
```zsh
ocobleseqx test %cat << EOF > /tmp/bgpadvs.yaml
apiVersion: metallb.io/v1beta1
kind: BGPAdvertisement
metadata:
  name: local
  namespace: metallb-system
  labels:
    cloud-provider: equinix-metal
spec:
  aggregationLength: 32
  localPref: 100
  communities:
  - 65535:65282
---
apiVersion: metallb.io/v1beta1
kind: BGPAdvertisement
metadata:
  name: external
  namespace: metallb-system
  labels:
    cloud-provider: equinix-metal
spec:
  aggregationLength: 24
EOF
kubectl apply -f /tmp/bgpadvs.yaml
bgpadvertisement.metallb.io/local created
bgpadvertisement.metallb.io/external created
ocobleseqx test %kubectl expose deployment nginx-hello-deployment --type=LoadBalancer --name=svc-1

service/svc-1 exposed
ocobleseqx test %kubectl -n metallb-system get ipaddresspools.metallb.io
kubectl -n metallb-system get bgppeers.metallb.io
No resources found in metallb-system namespace.
No resources found in metallb-system namespace.
ocobleseqx test %
ocobleseqx test %kubectl -n metallb-system get ipaddresspools.metallb.io
kubectl -n metallb-system get bgppeers.metallb.io
NAME            AUTO ASSIGN   AVOID BUGGY IPS   ADDRESSES
default.svc-1   false         false             ["147.28.154.166/32"]
NAME                                ADDRESS         ASN     BFD PROFILE   MULTI HOPS
metal-multiarch-k8s-x86-blue-00-0   169.254.255.1   65530                 
metal-multiarch-k8s-x86-blue-00-1   169.254.255.2   65530                 
metal-multiarch-k8s-x86-blue-01-0   169.254.255.1   65530                 
metal-multiarch-k8s-x86-blue-01-1   169.254.255.2   65530                 
ocobleseqx test %kubectl -n metallb-system get bgpadvertisements.metallb.io/equinix-metal-bgp-adv

Error from server (NotFound): bgpadvertisements.metallb.io "equinix-metal-bgp-adv" not found
ocobleseqx test %kubectl -n metallb-system get bgpadvertisements.metallb.io                      
NAME       IPADDRESSPOOLS   IPADDRESSPOOL SELECTORS   PEERS
external                                              
local                                                 
ocobleseqx test %kubectl delete svc svc-1

service "svc-1" deleted
ocobleseqx test %kubectl -n metallb-system get ipaddresspools.metallb.io
kubectl -n metallb-system get bgppeers.metallb.io
No resources found in metallb-system namespace.
No resources found in metallb-system namespace.
ocobleseqx test %kubectl -n metallb-system get bgpadvertisements.metallb.io

NAME       IPADDRESSPOOLS   IPADDRESSPOOL SELECTORS   PEERS
external                                              
local                                                 
ocobleseqx test %kubectl -n metallb-system delete bgpadvertisements.metallb.io local
kubectl -n metallb-system delete bgpadvertisements.metallb.io external
bgpadvertisement.metallb.io "local" deleted
bgpadvertisement.metallb.io "external" deleted
ocobleseqx test %
```

File: README.md [hunk 2]
```diff
@@ -382,13 +382,65 @@ If `kube-vip` management is enabled, then CCM does the following.
...
 
-**Supported Versions**: MetalLB [version 0.11.0](https://metallb.universe.tf/release-notes/#version-0-11-0) through [version 0.12.1](https://metallb.universe.tf/release-notes/#version-0-12-1). We are in the process of adding support for version 0.13.x.
+**Supported Versions**: MetalLB [version 0.11.0](https://metallb.universe.tf/release-notes/#version-0-11-0) through [version 0.13.7](https://metallb.universe.tf/release-notes/#version-0-13-7).
 
 When [MetalLB](https://metallb.universe.tf) is enabled, for user-deployed Kubernetes `Service` of `type=LoadBalancer`,
 the Equinix Metal CCM uses BGP and to provide the _equivalence_ of load balancing, without
 requiring an additional managed service (or hop). BGP route advertisements enable Equinix Metal's network
 to route traffic for your services at the Elastic IP to the correct host.
 
+**NOTE:** MetalLB [is now configurable via CRs](https://metallb.universe.tf/release-notes/#version-0-13-2), and ConfigMap configuration is not supported anymore. If you need to use a previous version, skip to the section [MetalLB from v0.11.0 to v0.12.1](#metallb-configuration-after-v0.13.2)
```
Comment (ctreatma) [inline README.md:392]: The second link is pointed at the wrong section (it goes to the 0.13.2+ instructions, but should go to the pre-0.13 instructions)

File: README.md [hunk 3]
```diff
@@ -382,13 +382,65 @@ If `kube-vip` management is enabled, then CCM does the following.
...
+* `metallb:///` - enable `MetalLB` management and update configuration in the default namespace `metallb-system`
+
+Notice the **three* slashes. In the URL, the namespace are in the path.
+
+When enabled, CCM controls the loadbalancer by updating the required MetalLB CR (custom resources).
+
+If `MetalLB` management is enabled, then CCM does the following.
+
+1. Get the appropriate namespace, based on the rules above.
+1.
```
Comment (ctreatma) [inline README.md:415]: Extra list item here

**Review by ctreatma — COMMENTED** [review]
A couple minor comments on the README.  If the DCO check is something we want to keep, could you make sure to sign off on the commits that are not passing the DCO check?

**Review by ctreatma — APPROVED** [review]

**Pull request merged**